### PR TITLE
Upgrades Tower to 0.5.1

### LINKED ISSUE #1569: upgrade to tower 0.5.0
violently failed ci from dependabot in https://github.com/kube-rs/kube/pull/1561

not sure if this is going to be hard or not. will have a look on the weekend to see if this worth doing before next release.

[tower 0.5.0 changes](https://github.com/tower-rs/tower/releases/tag/tower-0.5.0)

### PR DESCRIPTION
**Pull request by markdingram**
fixes #1569

## Solution

hyper-util 0.1.9 unblocked this change by dropping the tower 0.4 dep -> https://github.com/hyperium/hyper-util/pull/151

tower 0.5 includes this change to make Buffer generic over the Future rather than the Service -> https://github.com/tower-rs/tower/pull/654

tower 0.5 includes this change to rework Either to no longer change the error type to BoxError, meaning Kube-RS now has to do so itself -> https://github.com/tower-rs/tower/pull/637

tower-http 0.6.1 includes this change which promises to improve (de)compression -> https://github.com/tower-rs/tower-http/pull/521

### CODE REVIEW

File: kube-client/src/client/mod.rs [hunk 1]
```diff
@@ -75,8 +75,8 @@ pub use builder::{ClientBuilder, DynBody};
 #[derive(Clone)]
 pub struct Client {
     // - `Buffer` for cheap clone
-    // - `BoxService` for dynamic response future type
-    inner: Buffer<BoxService<Request<Body>, Response<Body>, BoxError>, Request<Body>>,
+    // - `BoxFuture` for dynamic response future type
+    inner: Buffer<Request<Body>, BoxFuture<'static, Result<Response<Body>, BoxError>>>,
```
Comment (clux) [inline kube-client/src/client/mod.rs:79]: This does seem like a nicer signature split. Interesting that it has to be 'static though. Is that a problem you think?

**Review by clux — APPROVED** [review]
Thanks a lot for this. From a quick pass this looks sensible to both me and CI.

File: kube-client/src/client/mod.rs (hunk 1 above)
Comment (markdingram) [reply kube-client/src/client/mod.rs:79]: `BoxService`can only be created by moving the underlying Service - https://docs.rs/tower/latest/tower/util/struct.BoxService.html#impl-BoxService%3CT,+U,+E%3E whereas a `BoxFuture` can be created from a [ref](https://docs.rs/futures/latest/futures/future/type.BoxFuture.html#impl-Pin%3CP%3E-1), hence generic over lifetime.

The `kube::Client` design owns the Buffer and to make Buffer `Clone` the Future needs to be `'static` - <https://tower-rs.github.io/tower/tower/buffer/struct.Buffer.html#impl-Clone-for-Buffer%3CReq,+F%3E>

So `'static` here is both unavoidable & fine (IMO!)

Comment (markdingram) [reply kube-client/src/client/mod.rs:79]: Similar in the Tonic upgrade PR - https://github.com/hyperium/tonic/pull/1892/files

Comment (clux) [reply kube-client/src/client/mod.rs:79]: Makes sense. Thanks for explaining!

**Pull request merged**